src/sync/discovery.ts

```diff
@@ -1,14 +1,93 @@
+import { getCloudInstallationId } from "../github/client/installation-id";
+import { DiscoveryMessagePayload } from "../sqs/discovery";
 
-export const DISCOVERY_LOGGER_NAME = "sync.discovery";
+/*
+* Mapping the response data into a map by repo.id as required by subscription.updateSyncState.
+*/
+const mapRepositories = (repositories: Repository[]): Repositories => {
+	return repositories.reduce((obj, repo) => {
+		obj[repo.id] = { repository: getRepositorySummary(repo) };
+		return obj;
+	}, {});
+};
+
+/*
+* Reset the sync count to zero.
+*/
+const resetSyncedReposCount = async (subscription: Subscription): Promise<void> => {
+	await subscription.updateSyncState({ numberOfSyncedRepos: 0 });
+};
+
+/*
+* Update the sync status of a batch of repos.
+*/
+const updateSyncState = async (subscription: Subscription, repositories: Repository[]): Promise<void> => {
+	const repos = mapRepositories(repositories);
+	await subscription.updateSyncState({
+		repos
+	});
+};
+
+/*
+* Checking the header atttribute link(type string[]) for a entry that contains "rel=next", this indicates more pages available.
+*/
+const checkHeaderForNextPage = (headers): boolean => {
+	const links = headers?.link?.split(",");
+	const regex = /(rel="next")/g;
+	const hasNextPage = links?.find(elem => elem.match(regex));
+	return hasNextPage;
+};
+
+/*
+* Continuosuly call the GitHub repo to fetch a page of repositories at a time and update there sync status until no more pages.
+*/
+const syncRepositories = async (github, subscription: Subscription, logger: LoggerWithTarget): Promise<void> => {
+	let page = 1;
+	let hasNextPage = true;
+	await resetSyncedReposCount(subscription);
+	while (hasNextPage) {
+		try {
+			const { data, headers } = await github.getRepositoriesPage(page);
+			await updateSyncState(subscription, data.repositories);
+			logger.info(`${data.repositories.length} Repositories syncing`);
+			hasNextPage = checkHeaderForNextPage(headers);
+			page++;
+		} catch (err) {
+			hasNextPage = false;
+			throw new Error(err);
+		}
+	}
+};
 
-export const discovery = (app: Application) => async (job, logger: LoggerWithTarget) => {
+/*
+* Use the github client to request all repositories and update the sync state per repo, send a bacnkfill queue message once complete.
+*/
+export const discovery = async (data: DiscoveryMessagePayload, logger: LoggerWithTarget): Promise<void> => {
+	const startTime = new Date();
+	const { jiraHost, installationId } = data;
+	const github = new GitHubClient(getCloudInstallationId(installationId), logger);
+	const subscription = await Subscription.getSingleInstallation(
+		jiraHost,
+		installationId
+	);
+
+	if(!subscription) {
+		logger.info({ jiraHost, installationId }, "Subscription has been removed, ignoring job.");
+		return;
+	}
+
+	await syncRepositories(github, subscription, logger);
+	await sqsQueues.backfill.sendMessage({ installationId, jiraHost, startTime: startTime.toISOString() }, 0, logger);
```

Comment:
you don't need to do `.toISOString()`, Date will serialize to it by default.  Just adding `startTime` should do.

Reply:
Did try that but the sendmessage BackfillMessagePayload type expects a string, I have moved it to its declaration though so its a bit cleaner passing it.